File: Solutions/102-BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.ts
import { TreeNode } from '../../Definitions/TreeNode/TreeNode';

function levelOrder(root: TreeNode | null): number[][] {
    const levelMap = new Map<number, number[]>();

    const levelTraversalHelper = (
        node: TreeNode | null,
        levelMap: Map<number, number[]>,
        level: number
    ) => {
        if (!node) return;

        if (node.left) {
            levelMap.set(level, [
                ...(levelMap.get(level) || []),
                node.left.val,
            ]);
            levelTraversalHelper(node.left, levelMap, level + 1);
        }

        if (node.right) {
            levelMap.set(level, [
                ...(levelMap.get(level) || []),
                node.right.val,
            ]);
            levelTraversalHelper(node.right, levelMap, level + 1);
        }
    };

    if (!root) return [];
    else levelMap.set(1, [root.val]);

    levelTraversalHelper(root, levelMap, 2);

    return Array.from(levelMap.values());
}

File: Solutions/102-BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.kt
class Solution {
    val levels: MutableList<MutableList<Int>> = mutableListOf()
    
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        dfs(root)
        return levels
    }

    fun dfs(node: TreeNode?, level: Int = 0){
        if(node == null) return

        if(levels.size === level) levels.add(mutableListOf())
        levels[level].add(node!!.`val`)
        
        dfs(node.left, level + 1)
        dfs(node.right, level + 1)
    }
}
